### PR TITLE
Flow rendering refactor

### LINKED ISSUE #1916: Flow visualization appearance enhancements
Add the following appearance enhancements to the Flow visualizer:

1. The ability to select draw integration paths as smooth tubes of user-specified diameter
2. The ability to include arrow heads indicating flow direction
3. The option to draw points (spheres) instead of lines
4. For pathlines, the ability to specify opacity based on age (length of the path line)

### LINKED ISSUE #2176: Flow line thickness
I would like to be able to adjust the line thickness for streamlines / flowlines. Thickness simply specified by an integer would be sufficient, however, maybe there is some applications where thickness defined by the value of some variable might be useful.

**Impact**
(*Please click check boxes AFTER submitting ticket*)
- [ ] High - User productivity significantly improved
- [x] Medium - User productivity partially improved
- [ ] Low - User productivity trivially improved

### PR DESCRIPTION
**Pull request by StasJ**
Fix #1916 
Fix #2176 

out2

### CODE REVIEW

**Comment by clyne** [comment]
GUI Review Comments:

1. Eliminate current Radius widget, which operates in world space,  and just have unit-less radius control with default of 1 approximating a 1 pixel line.

2. Reduce precision of radius selector to maybe two digits after decimal

3. Replace "stream" in Render Type selector with "curve" or "line"

4. Add a grouping widget to better indicate related capabilities. E.g. fade controls under Fade Flow tails

5. Add a less expensive sphere primitive that is less likely to blow up inexpensive graphics cards

6. Arbitrary limits on sliders such as "Every N Samples" should be changeable by user

7. Add controls for illumination equation when lighting is enabled (e.g. ambient, specular, diffuse). Be consistent with lighting controls used elsewhere. 

8. Remove "light from camera" option or make it the default

9. Support use case where user wants a single glyph displayed for each seed advected

10. Explore operating in units "time steps" instead of number of samples.

**Comment by StasJ** [comment]
2 and 6 are require refactoring of VWidgets so @sgpearse could you take a look at them if you have time.

2.) Currently the number of decimals is a global property of VLineEdit and VSliderEdit and there is no way to change for one specific case without refactoring VLineEdit and VSliderEdit. I think we could have a system where the number of decimal places is automatically made as short as possible unless the user manually specifies a higher precision value as this would eliminate such issues.

6.) This is a feature that needs to be added to VLineEdit.

File: share/shaders/FlowGlyphsLineDirArrow2D.frag [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#version 330 core
```
Comment (clyne) [inline share/shaders/FlowGlyphsLineDirArrow2D.frag:1]: Documentation

File: share/shaders/FlowGlyphsLineDirArrow2D.geom [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#version 330 core
```
Comment (clyne) [inline share/shaders/FlowGlyphsLineDirArrow2D.geom:1]: Documentation needed

File: share/shaders/FlowGlyphsSphere.frag [hunk 3]
```diff
@@ -0,0 +1,57 @@
+#version 330 core
```
Comment (clyne) [inline share/shaders/FlowGlyphsSphere.frag:1]: Document. What does this program do?

Comment (StasJ) [reply share/shaders/FlowGlyphsSphere.frag:1]: A bunch of these files that you requested documentation for are straightforward fragment shaders. These should be self-explanatory and thus none of the other (non raytracing) fragment shaders have documentation. Do you still want me to write documentation for these?

File: share/shaders/FlowGlyphsArrow.geom [hunk 4]
```diff
@@ -0,0 +1,112 @@
+#version 330 core
```
[the remark this answers is not in the document]
Comment (StasJ) [reply share/shaders/FlowGlyphsArrow.geom:11]: Done

File: share/shaders/FlowGlyphsArrow.geom [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+uniform float radius = 0.05;
+uniform int glyphStride = 1;
+uniform bool showOnlyLeadingSample = false;
+
+
+void EmitWorld(const vec3 v);
+void CylinderVert(const vec3 v, const vec3 d, const float value);
+
+
+void main()
```
[the remark this answers is not in the document]
Comment (StasJ) [reply share/shaders/FlowGlyphsArrow.geom:39]: done

File: share/shaders/FlowGlyphsArrow2D.frag [hunk 6]
```diff
@@ -0,0 +1,39 @@
+#version 330 core
```
[the remark this answers is not in the document]
Comment (StasJ) [reply share/shaders/FlowGlyphsArrow2D.frag:2]: Done

File: share/shaders/FlowGlyphsArrow2D.geom [hunk 7]
```diff
@@ -0,0 +1,63 @@
+#version 330 core
```
[the remark this answers is not in the document]
Comment (StasJ) [reply share/shaders/FlowGlyphsArrow2D.geom:5]: Done

File: share/shaders/FlowGlyphsLineDirArrow2D.frag (hunk 1 above)
Comment (StasJ) [reply share/shaders/FlowGlyphsLineDirArrow2D.frag:2]: Done

File: share/shaders/FlowGlyphsLineDirArrow2D.geom (hunk 2 above)
Comment (StasJ) [reply share/shaders/FlowGlyphsLineDirArrow2D.geom:3]: Done

File: share/shaders/FlowGlyphsSphere.frag (hunk 3 above)
Comment (StasJ) [reply share/shaders/FlowGlyphsSphere.frag]: Done

File: apps/vaporgui/FlowSubtabs.cpp [hunk 8]
```diff
@@ -119,9 +129,61 @@ void FlowVariablesSubtab::_dimensionalityChanged( int nDims ) const {
...
+    _pw = new PGroup;
 
     _layout->addWidget( _TFEditor, 0, 0 );
-
+    _layout->addWidget(_pw);
+    
+    PSection *ps;
+    
+    _pw->Add(ps = new PSection("Appearance"));
+    ps->Add(new PEnumDropdown(FlowParams::RenderTypeTag, {"Tubes", "Samples"}, {FlowParams::RenderTypeStream, FlowParams::RenderTypeSamples}, "Render Type"));
```
Comment (sgpearse) [inline apps/vaporgui/FlowSubtabs.cpp:140]: It looks like this will bring more low-level Params calls into the code base.  Do we want to fix this now or later?  It's hard to say now because the dust hasn't quite settled regarding PWidget usage.  I'd be ok with fixing this later, but I'm also inclined to steer away from this if we truly want the low-level interface to be inaccessible from the GUI.

File: apps/vaporgui/NavigationEventRouter.cpp [hunk 9]
```diff
@@ -34,6 +34,7 @@
 #include <qcheckbox.h>
 #include <QTextEdit>
 #include <QScrollArea>
+#include <cfloat>
```
Comment (sgpearse) [inline apps/vaporgui/NavigationEventRouter.cpp:37]: What does <cfloat> change in this file?

File: apps/vaporgui/VSubGroup.cpp [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#include "VSubGroup.h"
+#include <QVBoxLayout>
+
+VSubGroup::VSubGroup()
+{
+    QVBoxLayout *layout = new QVBoxLayout;
+    layout->setContentsMargins(12, 0, 0, 0);
```
Comment (sgpearse) [inline apps/vaporgui/VSubGroup.cpp:7]: This is a nicer solution than spaces :)

File: apps/vaporgui/PGroup.h [hunk 11]
```diff
@@ -19,5 +18,20 @@ class PGroup : public PWidget {
     PGroup *Add(PWidget *pw);
 
 protected:
+    QWidget *_widget;
```
Comment (sgpearse) [inline apps/vaporgui/PGroup.h:21]: Currently, all other PWidgets hold their QWidgets/VWidgets as private members.  Is there a reason that it is protected here, and private everywhere else?  Making them protected amongst all PWidgets would be more consistent, and allow all specialized PWidgets to be inheritable.

File: include/vapor/GLManager.h [hunk 12]
```diff
@@ -2,11 +2,12 @@
 
 #include "vapor/ShaderManager.h"
 #include "vapor/MatrixManager.h"
-#include "vapor/FontManager.h"
+//#include "vapor/FontManager.h"
```
Comment (sgpearse) [inline include/vapor/GLManager.h:5]: Can we delete?

File: lib/render/FlowRenderer.cpp [hunk 13]
```diff
@@ -322,17 +345,257 @@ FlowRenderer::_paintGL( bool fast )
...
+    
+    FlowParams::RenderType renderType = (FlowParams::RenderType)rp->GetValueLong(FlowParams::RenderTypeTag, FlowParams::RenderTypeStream);
+    FlowParams::GlpyhType glyphType = (FlowParams::GlpyhType)rp->GetValueLong(FlowParams::RenderGlyphTypeTag, FlowParams::GlpyhTypeSphere);
+    bool geom3d = rp->GetValueLong(FlowParams::RenderGeom3DTag, false);
+    bool show_dir = rp->GetValueLong(FlowParams::RenderShowStreamDirTag, false);
+    
+    float radiusBase = rp->GetValueDouble(FlowParams::RenderRadiusBaseTag, -1);
+    if (radiusBase == -1) {
+        vector<double> mind, maxd;
+        _dataMgr->GetVariableExtents(rp->GetCurrentTimestep(), rp->GetColorMapVariableName(), rp->GetRefinementLevel(), rp->GetCompressionLevel(), mind, maxd);
```
Comment (sgpearse) [inline lib/render/FlowRenderer.cpp:464]: I hope this doesn't come across as petty.  I've had difficulty in other parts of the code base because I need to resize my window, or reformat long lines to make sense of them.

A per-line character limit was documented in our Coding Convention, but we haven't been paying much attention to these rules.  I've recently violated some myself.  I think we should 1) try to enforce these, and 2) change them when needed.  

_7. Limit all lines to 80 characters whenever possible. Avoid deeply nested clauses and limit the number of lines in function/method to approximately one "page". Use two space indentation or tabs. Be consistent throughout the file.  
https://github.com/NCAR/VAPOR/wiki/Vapor-Coding-Convention_

An 80 character limit may be excessive.  This would mess with your header files that I think are nicely formatted, ie FlowRenderer.h.  A 100 character limit may strike a balance between making lines like this more readable, without requiring drastic changes to your code base.

File: apps/vaporgui/NavigationEventRouter.cpp (hunk 9 above)
Comment (StasJ) [reply apps/vaporgui/NavigationEventRouter.cpp:37]: It was being included at some point in the include tree but was no longer so I had to re-add it.

File: apps/vaporgui/PGroup.h (hunk 11 above)
Comment (StasJ) [reply apps/vaporgui/PGroup.h]: I made it private.

File: include/vapor/GLManager.h (hunk 12 above)
Comment (StasJ) [reply include/vapor/GLManager.h:5]: ok

File: apps/vaporgui/FlowSubtabs.cpp (hunk 8 above)
Comment (StasJ) [reply apps/vaporgui/FlowSubtabs.cpp:140]: Let's talk about this during the meeting

File: lib/render/FlowRenderer.cpp (hunk 13 above)
Comment (StasJ) [reply lib/render/FlowRenderer.cpp:464]: I think this is something that affects our entire code base. Let's talk about it during our meeting.

**Comment by shaomeng** [comment]
Is anyone able to compile this branch on cisl-vapor? It complains a QT problem, and I'm pretty I've put the latest 3rd party libraries on the systems directory. 

Thanks!

**Comment by clyne** [comment]
Works for me on my mac

> On Jun 3, 2020, at 12:50 PM, Samuel Li <notifications@github.com> wrote:
> 
> 
> Is anyone able to compile this branch on cisl-vapor? It complains a QT problem, and I'm pretty I've put the latest 3rd party libraries on the systems directory.
> 
> Thanks!
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

John Clyne
National Center for Atmospheric Research
303.497.1236 (w), 303.809.1922 (c)
clyne@ucar.edu

File: lib/render/FlowRenderer.cpp [hunk 14]
```diff
@@ -322,17 +345,257 @@ FlowRenderer::_paintGL( bool fast )
...
+    
+    if (params->GetValueLong("old_render", 0)) {
+        _renderFromAnAdvectionLegacy( &_advection, params, fast );
+        /* If the advection is bi-directional */
+        if( _2ndAdvection )
+            _renderFromAnAdvectionLegacy( _2ndAdvection.get(), params, fast );
+    } else {
+        // Workaround for how bi-directional was implemented.
+        if (_2ndAdvection)
+            _renderStatus = FlowStatus::SIMPLE_OUTOFDATE;
```
Comment (shaomeng) [inline lib/render/FlowRenderer.cpp:359]: Does this line duplicate with Line 364?

File: lib/render/FlowRenderer.cpp [hunk 15]
```diff
@@ -322,17 +345,257 @@ FlowRenderer::_paintGL( bool fast )
...
+    }
+    
     _restoreGLState();
 
+    return ret;
+}
+
+int FlowRenderer::_renderAdvection(const flow::Advection* adv)
+{
+    FlowParams *rp = (FlowParams*)GetActiveParams();
```
Comment (shaomeng) [inline lib/render/FlowRenderer.cpp:376]: Please use dynamic_cast for better safety.

File: lib/render/FlowRenderer.cpp [hunk 16]
```diff
@@ -322,17 +345,257 @@ FlowRenderer::_paintGL( bool fast )
...
+        vector<int> sizes;
+        vector<Vertex> sv;
+        
+        // If streams are larger than this then need to skip remaining
+        size_t maxSamples = rp->GetSteadyNumOfSteps() + 1;
+        
+        // First calculate the starting time stamp. Copied from legacy.
+        double startingTime;
+        if (!_cache_isSteady) {
+            int pastNumOfTimeSteps = ((FlowParams*)GetActiveParams())->GetPastNumOfTimeSteps();
```
Comment (shaomeng) [inline lib/render/FlowRenderer.cpp:392]: Please use dynamic_cast for better safety.

File: lib/render/FlowRenderer.cpp (hunk 13 above)
Comment (shaomeng) [reply lib/render/FlowRenderer.cpp:464]: I certainly think 80 character is a limit that's too small, thus I support increase it to a bigger value (100, 120, etc.).

At the same time I share the concern that a line that grows to an excessive length dramatically reduces its readability, and makes debugging difficult. 

As a side note, adopting tools such as clang-format and/or clang-tidy is a good solution for this type of hassles.

File: lib/render/FlowRenderer.cpp [hunk 17]
```diff
@@ -322,17 +345,257 @@ FlowRenderer::_paintGL( bool fast )
...
+        vec3 lens = max - min;
+        float largestDim  = glm::max(lens.x, glm::max(lens.y, lens.z));
+        radiusBase = largestDim / 560.f;
+        rp->SetValueDouble(FlowParams::RenderRadiusBaseTag, "", radiusBase);
+    }
+    float radiusScalar = rp->GetValueDouble(FlowParams::RenderRadiusScalarTag, 1);
+    float radius = radiusBase * radiusScalar;
+    int   glyphStride = rp->GetValueLong(FlowParams::RenderGlyphStrideTag, 5);
+    
+show_dir_loop:
```
Comment (shaomeng) [inline lib/render/FlowRenderer.cpp:476]: As goto is considered an uncommon logic flow, can you document what it does and why it is necessary? 

For example, my question is, Line 572-575 seem to indicate that variable `show_dir` is set to false right after show_dir_loop is executed. In that case, why not put show_dir_loop in a if conditioned branch?

**Comment by shaomeng** [comment]
Also, please make sure there are no compilation warnings and errors when compiling in release mode on cisl-vapor.

File: lib/render/FlowRenderer.cpp (hunk 16 above)
Comment (StasJ) [reply lib/render/FlowRenderer.cpp]: ok

File: lib/render/FlowRenderer.cpp (hunk 15 above)
Comment (StasJ) [reply lib/render/FlowRenderer.cpp]: ok

File: lib/render/FlowRenderer.cpp (hunk 14 above)
Comment (StasJ) [reply lib/render/FlowRenderer.cpp:359]: The value is reset by `_renderAdvection`

File: include/vapor/FlowParams.h [hunk 18]
```diff
@@ -25,13 +25,16 @@ enum class FlowDir : int
...
+};
 
 
 
 class PARAMS_API FlowParams : public RenderParams 
 {
 public:
+    
+    enum RenderType { RenderTypeStream, RenderTypeSamples};
+    enum GlpyhType { GlpyhTypeSphere, GlpyhTypeArrow };
```
[the remark this answers is not in the document]
Comment (StasJ) [reply include/vapor/FlowParams.h:37]: I did originally but it resulted in too many explicit casts.

File: lib/render/FlowRenderer.cpp (hunk 17 above)
Comment (StasJ) [reply lib/render/FlowRenderer.cpp:476]: It is just a different type of loop which repeats the rendering with show_dir disabled. I'm not sure what you mean by "put show_dir_loop in a if conditioned branch".

File: lib/render/FlowRenderer.cpp (hunk 13 above)
Comment (StasJ) [reply lib/render/FlowRenderer.cpp]: Done. Just thought I'd mention that you can configure your text editors to wrap at a certain character limit, optionally keeping words whole and text aligned with the current indentation. This way everyone can have their own preferred line width.

**Comment by shaomeng** [comment]
Hi Stas, are you sure the following warning/error is resolved on cisl-vapor? 

/home/shaomeng/Git/VAPOR-git-src/apps/vaporgui/PWidget.cpp: In member function ‘void PWidget::dynamicUpdateBegin()’:
/home/shaomeng/Git/VAPOR-git-src/apps/vaporgui/PWidget.cpp:131:30: error: ‘assert’ was not declared in this scope
     assert(_dynamicUpdateIsOn);


/home/shaomeng/Git/VAPOR-git-src/lib/render/FlowRenderer.cpp: In member function ‘int VAPoR::FlowRenderer::_renderAdvection(const flow::Advection*)’:
/home/shaomeng/Git/VAPOR-git-src/lib/render/FlowRenderer.cpp:435:25: warning: ‘startingTime’ may be used uninitialized in this function [-Wmaybe-uninitialized]
                         if(p.time >= startingTime)

File: lib/render/FlowRenderer.cpp (hunk 17 above)
Comment (shaomeng) [reply lib/render/FlowRenderer.cpp:482]: I meant if there's a reasonable alternative to `goto`? For example a regular if branch or things like that?

File: lib/render/FlowRenderer.cpp (hunk 14 above)
Comment (shaomeng) [reply lib/render/FlowRenderer.cpp:359]: Let me try again.

if you remove the current line 358 and 359, will the execution be different?

File: include/vapor/FlowParams.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (shaomeng) [reply include/vapor/FlowParams.h:37]: Can you elaborate a bit more? Why is an explicit cast required, and why a traditional enum result in less explicit casts?

File: lib/render/FlowRenderer.cpp (hunk 14 above)
Comment (StasJ) [reply lib/render/FlowRenderer.cpp:359]: Yes.

File: include/vapor/FlowParams.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (StasJ) [reply include/vapor/FlowParams.h:37]: Explicit casts are required because, like you said, enum classes result in stricter type safety. Traditional enums are essentially just constants.

File: lib/render/FlowRenderer.cpp (hunk 17 above)
Comment (StasJ) [reply lib/render/FlowRenderer.cpp:482]: Sure. All the goto does is reruns a portion of code with a different parameter in a special case. There are many ways to do this but I felt that the goto best represented the intention.

**Comment by StasJ** [comment]
```
cisl-vapor% make render
[ 10%] Built target common
[ 18%] Built target wasp
[ 41%] Built target vdc
[ 62%] Built target params
[ 66%] Built target flow
[ 68%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeRenderer.cpp.o
[ 68%] Building CXX object lib/render/CMakeFiles/render.dir/BarbRenderer.cpp.o
[ 68%] Building CXX object lib/render/CMakeFiles/render.dir/ControlExecutive.cpp.o
[ 68%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeIsoRenderer.cpp.o
[ 70%] Building CXX object lib/render/CMakeFiles/render.dir/Renderer.cpp.o
[ 70%] Building CXX object lib/render/CMakeFiles/render.dir/HelloRenderer.cpp.o
[ 70%] Building CXX object lib/render/CMakeFiles/render.dir/ShaderProgram.cpp.o
[ 70%] Building CXX object lib/render/CMakeFiles/render.dir/TwoDDataRenderer.cpp.o
[ 72%] Building CXX object lib/render/CMakeFiles/render.dir/TwoDRenderer.cpp.o
[ 72%] Building CXX object lib/render/CMakeFiles/render.dir/WireFrameRenderer.cpp.o
[ 72%] Building CXX object lib/render/CMakeFiles/render.dir/GeoTile.cpp.o
[ 75%] Building CXX object lib/render/CMakeFiles/render.dir/GeoTileMercator.cpp.o
[ 75%] Building CXX object lib/render/CMakeFiles/render.dir/GeoTileEquirectangular.cpp.o
[ 75%] Building CXX object lib/render/CMakeFiles/render.dir/GeoImage.cpp.o
[ 77%] Building CXX object lib/render/CMakeFiles/render.dir/GeoImageTMS.cpp.o
[ 77%] Building CXX object lib/render/CMakeFiles/render.dir/GeoImageGeoTiff.cpp.o
[ 79%] Building CXX object lib/render/CMakeFiles/render.dir/ImageRenderer.cpp.o
[ 79%] Building CXX object lib/render/CMakeFiles/render.dir/Visualizer.cpp.o
[ 79%] Building CXX object lib/render/CMakeFiles/render.dir/AnnotationRenderer.cpp.o
[ 79%] Building CXX object lib/render/CMakeFiles/render.dir/jfilewrite.cpp.o
[ 81%] Building CXX object lib/render/CMakeFiles/render.dir/RayCaster.cpp.o
[ 81%] Building CXX object lib/render/CMakeFiles/render.dir/FlowRenderer.cpp.o
[ 81%] Building CXX object lib/render/CMakeFiles/render.dir/ContourRenderer.cpp.o
[ 83%] Building CXX object lib/render/CMakeFiles/render.dir/SliceRenderer.cpp.o
[ 83%] Building CXX object lib/render/CMakeFiles/render.dir/MyPython.cpp.o
[ 83%] Building CXX object lib/render/CMakeFiles/render.dir/MatrixManager.cpp.o
[ 85%] Building CXX object lib/render/CMakeFiles/render.dir/Shader.cpp.o
[ 85%] Building CXX object lib/render/CMakeFiles/render.dir/LegacyGL.cpp.o
[ 85%] Building CXX object lib/render/CMakeFiles/render.dir/ShaderManager.cpp.o
[ 87%] Building CXX object lib/render/CMakeFiles/render.dir/GLManager.cpp.o
[ 87%] Building CXX object lib/render/CMakeFiles/render.dir/FontManager.cpp.o
[ 87%] Building CXX object lib/render/CMakeFiles/render.dir/Font.cpp.o
[ 87%] Building CXX object lib/render/CMakeFiles/render.dir/TextLabel.cpp.o
[ 89%] Building CXX object lib/render/CMakeFiles/render.dir/PyEngine.cpp.o
[ 89%] Building CXX object lib/render/CMakeFiles/render.dir/CalcEngineMgr.cpp.o
[ 89%] Building CXX object lib/render/CMakeFiles/render.dir/GeoTIFWriter.cpp.o
[ 91%] Building CXX object lib/render/CMakeFiles/render.dir/ImageWriter.cpp.o
[ 91%] Building CXX object lib/render/CMakeFiles/render.dir/JPGWriter.cpp.o
[ 91%] Building CXX object lib/render/CMakeFiles/render.dir/PNGWriter.cpp.o
[ 93%] Building CXX object lib/render/CMakeFiles/render.dir/TIFWriter.cpp.o
[ 93%] Building CXX object lib/render/CMakeFiles/render.dir/Proj4StringParser.cpp.o
[ 93%] Building CXX object lib/render/CMakeFiles/render.dir/glutil.cpp.o
[ 95%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeAlgorithm.cpp.o
[ 95%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeRegular.cpp.o
[ 95%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeCellTraversal.cpp.o
[ 97%] Building CXX object lib/render/CMakeFiles/render.dir/Texture.cpp.o
[ 97%] Building CXX object lib/render/CMakeFiles/render.dir/Framebuffer.cpp.o
[ 97%] Building CXX object lib/render/CMakeFiles/render.dir/ModelRenderer.cpp.o
[100%] Linking CXX shared library ../librender.so
[100%] Built target render
cisl-vapor%
```

File: include/vapor/FlowParams.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (shaomeng) [reply include/vapor/FlowParams.h:37]: Can you elaborate why you will need too many explicit casts, no matter new or old style enums? 

Regarding safety, it's exactly because old-style enums permit an implicit cast, thus there's more chance the program goes wrong without the awareness of the programmer, isn't it? So the new-style enum is safer because its usage is more limited, isn't it?

File: lib/render/FlowRenderer.cpp (hunk 14 above)
Comment (shaomeng) [reply lib/render/FlowRenderer.cpp:359]: Ugh, if it were me I'd put a better explanation than a single-worded "Yes."

Overall, it's more of your interest than anyone else's interest to have this PR resolved and merged, isn't it?

File: lib/render/FlowRenderer.cpp (hunk 17 above)
Comment (shaomeng) [reply lib/render/FlowRenderer.cpp:482]: In that case, are you able to adopt a conditional instead of a goto? 

The main reason is that goto isn't very commonly used, and it certainly doesn't make the code easier to understand. There are very few cases goto is considered a better choice than others, but I'm afraid this isn't one of those cases.

More comprehensive arguments against using goto is available on the internet, which I'm sure you can easily find.

Comment (clyne) [reply lib/render/FlowRenderer.cpp:482]: I'm with Sam on this one. The control flow is not particularly clear. Couldn't you simply put code after the label in a helper function?

File: lib/render/FlowRenderer.cpp (hunk 14 above)
Comment (clyne) [reply lib/render/FlowRenderer.cpp:359]: A simple comment that _renderAdvection resets _2ndAdvection would clarify what is happening here.

Comment (StasJ) [reply lib/render/FlowRenderer.cpp:363]: Done.

File: lib/render/FlowRenderer.cpp (hunk 17 above)
Comment (StasJ) [reply lib/render/FlowRenderer.cpp]: Done.

File: lib/render/FlowRenderer.cpp (hunk 14 above)
Comment (shaomeng) [reply lib/render/FlowRenderer.cpp:363]: @clyne It doesn't seem your description is accurate. The function `_renderAdvection()` takes in an advection pointed by a const pointer, and renders whatever that Advection provides. How can a function reset a const pointer?

Comment (clyne) [reply lib/render/FlowRenderer.cpp:363]: @shaomeng see comments added by Stas'

Comment (shaomeng) [reply lib/render/FlowRenderer.cpp:363]: @clyne I saw that but I don't think my question is answered in either comments or this conversation list. Let me try again: 

In the current code, line 362 and 367 perform the same test, but I don't see anything could possibly change between these two lines that results in different test results. This makes me think if removing the test at line 362 would actually change the execution from the current form. If the answer is yes, I'd appreciated a more detailed explanation on where and under what condition the execution would change.

Comment (clyne) [reply lib/render/FlowRenderer.cpp:363]: It's not the test results that will change, it's the behavior of _renderAdvection that will change because it references  _renderStatus

Comment (shaomeng) [reply lib/render/FlowRenderer.cpp:363]: Thanks for the explanation John. It makes more sense now, but the logic still seems confusing: the rendering of the first advection (line 365) is dependent on if the second advection exists or not (line 362-363) ?? 

Stas, I'm curious why it is designed this way?

**Comment by shaomeng** [comment]
> I just tested it on cisl-vapor and it compiles without warnings. Also it passed the warning test on our ubuntu-16 test machine as indicated by the checkmark.
> 
> ```
> cisl-vapor% make render
> [ 10%] Built target common
> [ 18%] Built target wasp
> [ 41%] Built target vdc
> [ 62%] Built target params
> [ 66%] Built target flow
> [ 68%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeRenderer.cpp.o
> [ 68%] Building CXX object lib/render/CMakeFiles/render.dir/BarbRenderer.cpp.o
> [ 68%] Building CXX object lib/render/CMakeFiles/render.dir/ControlExecutive.cpp.o
> [ 68%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeIsoRenderer.cpp.o
> [ 70%] Building CXX object lib/render/CMakeFiles/render.dir/Renderer.cpp.o
> [ 70%] Building CXX object lib/render/CMakeFiles/render.dir/HelloRenderer.cpp.o
> [ 70%] Building CXX object lib/render/CMakeFiles/render.dir/ShaderProgram.cpp.o
> [ 70%] Building CXX object lib/render/CMakeFiles/render.dir/TwoDDataRenderer.cpp.o
> [ 72%] Building CXX object lib/render/CMakeFiles/render.dir/TwoDRenderer.cpp.o
> [ 72%] Building CXX object lib/render/CMakeFiles/render.dir/WireFrameRenderer.cpp.o
> [ 72%] Building CXX object lib/render/CMakeFiles/render.dir/GeoTile.cpp.o
> [ 75%] Building CXX object lib/render/CMakeFiles/render.dir/GeoTileMercator.cpp.o
> [ 75%] Building CXX object lib/render/CMakeFiles/render.dir/GeoTileEquirectangular.cpp.o
> [ 75%] Building CXX object lib/render/CMakeFiles/render.dir/GeoImage.cpp.o
> [ 77%] Building CXX object lib/render/CMakeFiles/render.dir/GeoImageTMS.cpp.o
> [ 77%] Building CXX object lib/render/CMakeFiles/render.dir/GeoImageGeoTiff.cpp.o
> [ 79%] Building CXX object lib/render/CMakeFiles/render.dir/ImageRenderer.cpp.o
> [ 79%] Building CXX object lib/render/CMakeFiles/render.dir/Visualizer.cpp.o
> [ 79%] Building CXX object lib/render/CMakeFiles/render.dir/AnnotationRenderer.cpp.o
> [ 79%] Building CXX object lib/render/CMakeFiles/render.dir/jfilewrite.cpp.o
> [ 81%] Building CXX object lib/render/CMakeFiles/render.dir/RayCaster.cpp.o
> [ 81%] Building CXX object lib/render/CMakeFiles/render.dir/FlowRenderer.cpp.o
> [ 81%] Building CXX object lib/render/CMakeFiles/render.dir/ContourRenderer.cpp.o
> [ 83%] Building CXX object lib/render/CMakeFiles/render.dir/SliceRenderer.cpp.o
> [ 83%] Building CXX object lib/render/CMakeFiles/render.dir/MyPython.cpp.o
> [ 83%] Building CXX object lib/render/CMakeFiles/render.dir/MatrixManager.cpp.o
> [ 85%] Building CXX object lib/render/CMakeFiles/render.dir/Shader.cpp.o
> [ 85%] Building CXX object lib/render/CMakeFiles/render.dir/LegacyGL.cpp.o
> [ 85%] Building CXX object lib/render/CMakeFiles/render.dir/ShaderManager.cpp.o
> [ 87%] Building CXX object lib/render/CMakeFiles/render.dir/GLManager.cpp.o
> [ 87%] Building CXX object lib/render/CMakeFiles/render.dir/FontManager.cpp.o
> [ 87%] Building CXX object lib/render/CMakeFiles/render.dir/Font.cpp.o
> [ 87%] Building CXX object lib/render/CMakeFiles/render.dir/TextLabel.cpp.o
> [ 89%] Building CXX object lib/render/CMakeFiles/render.dir/PyEngine.cpp.o
> [ 89%] Building CXX object lib/render/CMakeFiles/render.dir/CalcEngineMgr.cpp.o
> [ 89%] Building CXX object lib/render/CMakeFiles/render.dir/GeoTIFWriter.cpp.o
> [ 91%] Building CXX object lib/render/CMakeFiles/render.dir/ImageWriter.cpp.o
> [ 91%] Building CXX object lib/render/CMakeFiles/render.dir/JPGWriter.cpp.o
> [ 91%] Building CXX object lib/render/CMakeFiles/render.dir/PNGWriter.cpp.o
> [ 93%] Building CXX object lib/render/CMakeFiles/render.dir/TIFWriter.cpp.o
> [ 93%] Building CXX object lib/render/CMakeFiles/render.dir/Proj4StringParser.cpp.o
> [ 93%] Building CXX object lib/render/CMakeFiles/render.dir/glutil.cpp.o
> [ 95%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeAlgorithm.cpp.o
> [ 95%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeRegular.cpp.o
> [ 95%] Building CXX object lib/render/CMakeFiles/render.dir/VolumeCellTraversal.cpp.o
> [ 97%] Building CXX object lib/render/CMakeFiles/render.dir/Texture.cpp.o
> [ 97%] Building CXX object lib/render/CMakeFiles/render.dir/Framebuffer.cpp.o
> [ 97%] Building CXX object lib/render/CMakeFiles/render.dir/ModelRenderer.cpp.o
> [100%] Linking CXX shared library ../librender.so
> [100%] Built target render
> cisl-vapor%
> ```

Stas, did you try building the entire project instead of a specific object?
